### PR TITLE
Multicategory sorting fixes

### PR DESCRIPTION
**Pull request by etpinard**
fixes https://github.com/plotly/plotly.js/issues/3356 - and made me realised that `multicategory` are slightly more complicated than first expected. 

In brief, we need to (1) scan the data array (2) sort the resulting multi-category list (3) loop over the data arrays again to retrieve their `calcdata` values.

cc @plotly/plotly_js

### CODE REVIEW

File: src/plots/plots.js [hunk 1]
```diff
@@ -2617,7 +2617,7 @@ plots.doCalcdata = function(gd, traces) {
     }
 
     // clear stuff that should recomputed in 'regular' loop
-    if(hasCalcTransform) clearAxesCalc(axList);
+    if(hasCalcTransform) setupAxisCategories(axList);
```
Comment (alexcjohnson) [inline src/plots/plots.js]: `, fullData`?

File: src/plots/plots.js [hunk 2]
```diff
@@ -2675,9 +2675,13 @@ plots.doCalcdata = function(gd, traces) {
...
 
-function clearAxesCalc(axList) {
+function setupAxisCategories(axList, fullData) {
     for(var i = 0; i < axList.length; i++) {
-        axList[i].clearCalc();
+        var ax = axList[i];
+        ax.clearCalc();
+        if(ax.type === 'multicategory') {
+            ax.setupMultiCategory(fullData);
+        }
```
Comment (alexcjohnson) [inline src/plots/plots.js]: `clearCalc` is also used in `rangeslider.draw` and `gl2d/scene2d` - do these need `setupMultiCategory`? Should that just be part of `clearCalc` for multicategory axes?

Comment (etpinard) [reply src/plots/plots.js]: > `clearCalc` is also used in `rangeslider.draw` and `gl2d/scene2d` - do these need `setupMultiCategory`?

`setupMultiCategory` only needs to be called once (here in `doCalcdata`, to fill in `ax._muticatList`), so rangeslider don't need it, that's why I didn't put the `setupMultiCategory` logic in `clearCalc`. As for `gl2d/scene2d`, gl2d subplots don't support `multicategory` axes, so I think we're :ok_hand:

File: src/plots/plots.js (hunk 1 above)
Comment (alexcjohnson) [reply src/plots/plots.js]: maybe a can of worms, but is there a test for multicat + calctransforms (or non-calc-transforms for that matter)?

Comment (etpinard) [reply src/plots/plots.js]: Multicategory axes don't support transforms. I'm not sure how gracefully things fail though.

Comment (alexcjohnson) [reply src/plots/plots.js]: OK, that's fine. Nothing more from me - I'll let one of the other folks give this a once-over and the 💃

Comment (etpinard) [reply src/plots/plots.js]: Transforms + multicategory tracked in -> https://github.com/plotly/plotly.js/issues/3363

File: test/image/mocks/multicategory-sorting.json [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        "yaxis": "y2"
+    },
+
+    {
+        "x": [[1, 2, 3, 4, 5, 6, 4, 5, 6, 7, 8, 9],
+              [1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2]],
+        "y":  [1, 2, 3, 4, 5, 6, 1, 2, 3, 4, 5, 6],
+        "xaxis": "x3",
+        "yaxis": "y3"
+    }],
```
Comment (archmoj) [inline test/image/mocks/multicategory-sorting.json]: @etpinard Would you please adding these two traces to the end of this mock? I was curious about the `null` cases. But your algorithm was able to ignore them. 
<pre>
{
        "x": [[1, 2, 3, 4, 5, 6],
              [1, 2, 1, 2, null, 2]],
        "y":  [1, 2, 3, 4, 5, 6],
        "xaxis": "x4",
        "yaxis": "y4"
    },
    {
        "x": [[4, 5, 6, 7, 8, 9],
              [1, null, 1, 2, 1, 2]],
        "y":  [1, 2, 3, 4, 5, 6],
        "xaxis": "x4",
        "yaxis": "y4"
    }
</pre>

Comment (etpinard) [reply test/image/mocks/multicategory-sorting.json]: Good call! Done in -> 96edd7215

**Comment by archmoj** [comment]
:dancer:

**Pull request merged**